util/mkpod2html.pl

```diff
@@ -0,0 +1,52 @@
+#! /usr/bin/env perl
+# Copyright 2020 The OpenSSL Project Authors. All Rights Reserved.
+#
+# Licensed under the Apache License 2.0 (the "License").  You may not use
+# this file except in compliance with the License.  You can obtain a copy
+# in the file LICENSE in the source distribution or at
+# https://www.openssl.org/source/license.html
+
+use strict;
+use warnings;
+
+use lib ".";
+use Getopt::Std;
+use Pod::Html;
+
+# Options.
+our($opt_i);    # -i INFILE
+our($opt_o);    # -o OUTFILE
+our($opt_t);    # -t TITLE
+our($opt_r);    # -r PODROOT
+
+getopts('i:o:t:r:');
+die "-i flag missing" unless $opt_i;
+die "-o flag missing" unless $opt_o;
+die "-t flag missing" unless $opt_t;
+die "-r flag missing" unless $opt_r;
+
+pod2html
+    "--infile=$opt_i",
+    "--outfile=$opt_o",
+    "--title=$opt_t",
+    "--podroot=$opt_r",
+    "--podpath=man1:man3:man5:man7",
+    "--htmldir=..";
+
+# Read in contents.
+open F, "<$opt_o"
+    or die "Can't read $opt_o, $!";
+my $contents = '';
+{
+    local $/ = undef;
+    $contents = <F>;
+}
+close F;
+unlink $opt_o;
+
+$contents =~
+    s|href="http://man\.he\.net/(man\d/[^"]+)(?:\.html)?"|href="../$1.html|g;
```

Comment:
The substitution regex looks a lot better now after moving it into a perl file. But while reviewing the pr I noticed that this improvement actually just reverts 'complications' introduced by commit b0940b33a6:

The regex _used to be_ in a perl file,  `process_docs.pl`

https://github.com/openssl/openssl/blob/22db21099dd89f4b936137410a3193aebbc189f7/util/process_docs.pl#L122

until b0940b33a6 moved the code into the makefile templates. (*Note:* `process_docs.pl` is still there, but doesn't seem to be used anymore.)

This move came at the price of tripling the regex and of heavy regex quoting, because the regex is evaluated multiple times (template -> makefile -> shell/cmd.exe -> perl, see my comment in commit message a9e4e3c39e).

https://github.com/openssl/openssl/blob/c8a557357789a8a23fe7b87ec352bd5ea2c1251b/Configurations/descrip.mms.tmpl#L790
https://github.com/openssl/openssl/blob/c8a557357789a8a23fe7b87ec352bd5ea2c1251b/Configurations/unix-Makefile.tmpl#L1192
https://github.com/openssl/openssl/blob/c8a557357789a8a23fe7b87ec352bd5ea2c1251b/Configurations/windows-makefile.tmpl#L594


Now the code goes back into a perl file, albeit a different one, `mkpod2html.pl`. 

https://github.com/openssl/openssl/blob/22db21099dd89f4b936137410a3193aebbc189f7/util/mkpod2html.pl#L48

Why was `process_docs.pl` abandoned in the first place? I don't really understand the benefit of the changes in b0940b33a6. My personal impression is that the logic got more complicated but the speed didn't improve (I might be wrong, I didn't do any measurements). Also it confused me when reviewing #10132 that some `*.pod.in` files seem to be generated at build time while others are generated at installation time.

Maybe I am just missing the grand picture @levitte?

Reply:
@mspncp: what pod.in files are processed at install time?  If so, that's a bug -- all manpages should be generated as part of the ```build_generated``` target.

Reply:
`process-docs.pl` was kind of a magic wand that did a little too much, and not always well.  It seemed better to use the actual commands for each case appropriately.  `mkpod2html.pl` is more of an almost drop in replacement for the normal pod2html.  Why I didn't remove it when I stopped its usage, I dunno.

Reply:
> @mspncp: what pod.in files are processed at install time? If so, that's a bug -- all manpages should be generated as part of the `build_generated` target.


I ran this command

```bash
$ find doc \( -name  '*.pod' -o -name '*.pod.in' \) | sort
```

three times on the checkout  31dc2ac6f63423be359da6985e57e72dd4d762a1 to obtain a sorted list of the files:

- 01-before       (after make clean)
- 02-after-make
- 03-after-make-install

Here are the changes:

#### After make

```diff
--- 01-before	2020-01-18 08:45:02.189862195 +0100
+++ 02-after-make	2020-01-18 08:48:05.961821622 +0100
@@ -24,57 +24,84 @@
 doc/internal/man7/build.info.pod
 doc/man1/CA.pl.pod
 doc/man1/openssl-asn1parse.pod
+doc/man1/openssl-ca.pod
 doc/man1/openssl-ca.pod.in
 doc/man1/openssl-ciphers.pod
 doc/man1/openssl-cmds.pod
+doc/man1/openssl-cms.pod
 doc/man1/openssl-cms.pod.in
+doc/man1/openssl-crl.pod
 doc/man1/openssl-crl.pod.in
 doc/man1/openssl-crl2pkcs7.pod
+doc/man1/openssl-dgst.pod
 doc/man1/openssl-dgst.pod.in
+doc/man1/openssl-dhparam.pod
 doc/man1/openssl-dhparam.pod.in
 doc/man1/openssl-dsa.pod.in
+doc/man1/openssl-dsaparam.pod
 doc/man1/openssl-dsaparam.pod.in
 doc/man1/openssl-ec.pod.in
+doc/man1/openssl-ecparam.pod
 doc/man1/openssl-ecparam.pod.in
+doc/man1/openssl-enc.pod
 doc/man1/openssl-enc.pod.in
 doc/man1/openssl-engine.pod
 doc/man1/openssl-errstr.pod
 doc/man1/openssl-fipsinstall.pod
+doc/man1/openssl-gendsa.pod
 doc/man1/openssl-gendsa.pod.in
 doc/man1/openssl-genpkey.pod.in
+doc/man1/openssl-genrsa.pod
 doc/man1/openssl-genrsa.pod.in
 doc/man1/openssl-info.pod
 doc/man1/openssl-kdf.pod
 doc/man1/openssl-list.pod
 doc/man1/openssl-mac.pod
 doc/man1/openssl-nseq.pod
+doc/man1/openssl-ocsp.pod
 doc/man1/openssl-ocsp.pod.in
+doc/man1/openssl-passwd.pod
 doc/man1/openssl-passwd.pod.in
+doc/man1/openssl-pkcs12.pod
 doc/man1/openssl-pkcs12.pod.in
 doc/man1/openssl-pkcs7.pod.in
+doc/man1/openssl-pkcs8.pod
 doc/man1/openssl-pkcs8.pod.in
 doc/man1/openssl-pkey.pod.in
 doc/man1/openssl-pkeyparam.pod.in
+doc/man1/openssl-pkeyutl.pod
 doc/man1/openssl-pkeyutl.pod.in
 doc/man1/openssl-prime.pod
 doc/man1/openssl-provider.pod
+doc/man1/openssl-rand.pod
 doc/man1/openssl-rand.pod.in
 doc/man1/openssl-rehash.pod
+doc/man1/openssl-req.pod
 doc/man1/openssl-req.pod.in
 doc/man1/openssl-rsa.pod.in
+doc/man1/openssl-rsautl.pod
 doc/man1/openssl-rsautl.pod.in
+doc/man1/openssl-s_client.pod
 doc/man1/openssl-s_client.pod.in
+doc/man1/openssl-s_server.pod
 doc/man1/openssl-s_server.pod.in
+doc/man1/openssl-s_time.pod
 doc/man1/openssl-s_time.pod.in
 doc/man1/openssl-sess_id.pod
+doc/man1/openssl-smime.pod
 doc/man1/openssl-smime.pod.in
+doc/man1/openssl-speed.pod
 doc/man1/openssl-speed.pod.in
 doc/man1/openssl-spkac.pod.in
+doc/man1/openssl-srp.pod
 doc/man1/openssl-srp.pod.in
 doc/man1/openssl-storeutl.pod.in
+doc/man1/openssl-ts.pod
 doc/man1/openssl-ts.pod.in
+doc/man1/openssl-verify.pod
 doc/man1/openssl-verify.pod.in
 doc/man1/openssl-version.pod
+doc/man1/openssl-x509.pod
 doc/man1/openssl-x509.pod.in
 doc/man1/openssl.pod
 doc/man1/tsget.pod
@@ -621,6 +648,7 @@
 doc/man7/evp.pod
 doc/man7/openssl-core.h.pod
 doc/man7/openssl-env.pod
+doc/man7/openssl_user_macros.pod
 doc/man7/openssl_user_macros.pod.in
 doc/man7/ossl_store-file.pod
 doc/man7/ossl_store.pod
```

#### After make install

```diff
--- 02-after-make	2020-01-18 08:48:05.961821622 +0100
+++ 03-after-make-install	2020-01-18 08:49:52.286955242 +0100
@@ -37,9 +37,11 @@
 doc/man1/openssl-dgst.pod.in
 doc/man1/openssl-dhparam.pod
 doc/man1/openssl-dhparam.pod.in
+doc/man1/openssl-dsa.pod
 doc/man1/openssl-dsa.pod.in
 doc/man1/openssl-dsaparam.pod
 doc/man1/openssl-dsaparam.pod.in
+doc/man1/openssl-ec.pod
 doc/man1/openssl-ec.pod.in
 doc/man1/openssl-ecparam.pod
 doc/man1/openssl-ecparam.pod.in
@@ -50,6 +52,7 @@
 doc/man1/openssl-fipsinstall.pod
 doc/man1/openssl-gendsa.pod
 doc/man1/openssl-gendsa.pod.in
+doc/man1/openssl-genpkey.pod
 doc/man1/openssl-genpkey.pod.in
 doc/man1/openssl-genrsa.pod
 doc/man1/openssl-genrsa.pod.in
@@ -64,10 +67,13 @@
 doc/man1/openssl-passwd.pod.in
 doc/man1/openssl-pkcs12.pod
 doc/man1/openssl-pkcs12.pod.in
+doc/man1/openssl-pkcs7.pod
 doc/man1/openssl-pkcs7.pod.in
 doc/man1/openssl-pkcs8.pod
 doc/man1/openssl-pkcs8.pod.in
+doc/man1/openssl-pkey.pod
 doc/man1/openssl-pkey.pod.in
+doc/man1/openssl-pkeyparam.pod
 doc/man1/openssl-pkeyparam.pod.in
 doc/man1/openssl-pkeyutl.pod
 doc/man1/openssl-pkeyutl.pod.in
@@ -78,6 +84,7 @@
 doc/man1/openssl-rehash.pod
 doc/man1/openssl-req.pod
 doc/man1/openssl-req.pod.in
+doc/man1/openssl-rsa.pod
 doc/man1/openssl-rsa.pod.in
 doc/man1/openssl-rsautl.pod
 doc/man1/openssl-rsautl.pod.in
@@ -92,9 +99,11 @@
 doc/man1/openssl-smime.pod.in
 doc/man1/openssl-speed.pod
 doc/man1/openssl-speed.pod.in
+doc/man1/openssl-spkac.pod
 doc/man1/openssl-spkac.pod.in
 doc/man1/openssl-srp.pod
 doc/man1/openssl-srp.pod.in
+doc/man1/openssl-storeutl.pod
 doc/man1/openssl-storeutl.pod.in
 doc/man1/openssl-ts.pod
 doc/man1/openssl-ts.pod.in
```